src/hometask8/Main.java

```diff
@@ -0,0 +1,99 @@
+package hometask8;
+
+
+import java.time.Year;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class Main {
+    public static void main(String[] args) throws InterruptedException {
+        ArrayList<Product> products = new ArrayList<>();
+        products.add(new Product("Book", 300, 2));
+        Thread.sleep(1000);
+        products.add(new Product("Book", 10, 10));
+        Thread.sleep(1000);
+        products.add(new Product("Bike", 200, 15));
+        Thread.sleep(1000);
+        products.add(new Product("Book", 150, 7));
+        Thread.sleep(1000);
+        products.add(new Product("Car", 250, 3));
+        Thread.sleep(1000);
+        products.add(new Product("Book", 30, 4));
+        Thread.sleep(1000);
+        products.add(new Product("toy", 250, 20));
+        Thread.sleep(1000);
+        products.add(new Product("Book", 1000, 20));
+        products.add(new Product("Book", 500, 10));
+
+//        1.2
+//        findBooks1(products);
+//        2.2
+//        findBooksAndApplyDiscount(products);
+//        3.3
+//        findCheapestBook(products);
+//        4.2
+//        findThreeLatestProducts(products);
+//        5.2
+//        sumOfCertainProducts(products);
+//        6.2
+//        groupProducts(products);
+    }
+
+    // Method for 1.2
+    public static void findBooks1(ArrayList<Product> products) {
+        ArrayList<Product> foundBooks = products.stream()
+                .filter(item -> (item.getType().equals("Book") && item.getPrice() > 250))
+                .collect(Collectors.toCollection(ArrayList::new));
+        System.out.println(foundBooks);
+    }
+
+    // Method for 2.2
+    public static void findBooksAndApplyDiscount(ArrayList<Product> products) {
+        ArrayList<Product> foundBooks = products.stream()
+                .filter(item -> (item.getType().equals("Book")))
+                .peek(item -> item.setDiscount(10))
```

Comment:
There is no actual calculation and setting discounted value. Consider using Product.discount as boolean and calculate the new value based on discount applicability like:
.filter(Product:: discount)
.map(product -> {
  product.setPrice(produc.getPrice() * 0.9);
  return product;
}

Reply:
But it calculates price by itself in setDiscount
When I set new Discount value the price calculates also